--- a/drivers/misc/lowmemnotify.c
+++ b/drivers/misc/lowmemnotify.c
@@ -194,7 +194,7 @@
 	if (other_free > totalreserve_pages)
 		free += other_free - totalreserve_pages;
 
-	return free;
+	return free + nr_swap_pages;
 }
 EXPORT_SYMBOL(memnotify_get_free);
 
@@ -206,15 +206,13 @@
 unsigned long memnotify_get_used(void)
 {
 	unsigned long used_mem;
-	unsigned long used_swap;
 
 	unsigned long free_mem;
 	
 	free_mem = memnotify_get_free();
-	used_swap = total_swap_pages - nr_swap_pages;
-	used_mem = totalram_pages - free_mem;
+	used_mem = totalram_pages + total_swap_pages - free_mem;
 
-	return used_mem + used_swap;
+	return used_mem;
 }
 
 /** 
@@ -231,7 +229,7 @@
 	int i;
 
 	used = memnotify_get_used();
-	used_ratio = used * 100 / totalram_pages;
+	used_ratio = used * 100 / (totalram_pages + total_swap_pages);
 
 	threshold = THRESHOLD_NORMAL;
 	last_threshold = atomic_read(&memnotify_last_threshold);
@@ -351,7 +349,7 @@
 	int i;
 
 	used = memnotify_get_used();
-	total_mem = totalram_pages;
+	total_mem = totalram_pages + total_swap_pages;
 
 	threshold = memnotify_threshold();
 	last_threshold = atomic_read(&memnotify_last_threshold);
@@ -362,7 +360,7 @@
 		"Used (Mem+Swap): %ldMB\n", MB(used));
 
 	len += snprintf(buf+len, PAGE_SIZE,
-		"Used (Mem): %ldMB\n", MB(totalram_pages-memnotify_get_free()));
+		"Used (Mem): %ldMB\n", MB(totalram_pages + nr_swap_pages - memnotify_get_free()));
 
 	len += snprintf(buf+len, PAGE_SIZE,
 		"Used (Swap): %ldMB\n", MB(total_swap_pages - nr_swap_pages));
