--- linux-2.6.24/arch/arm/mach-omap3pe/prcm_clk.c	2010-04-22 21:58:25.000000000 +0000
+++ linux-2.6.24_sbromwich/arch/arm/mach-omap3pe/prcm_clk.c	2010-04-19 22:24:12.000000000 +0000
@@ -88,7 +88,7 @@
 		udelay(5);
 	} while (retries--);
 
-	printk(KERN_INFO "Timeout in check_device_status(), deviceid %d, "
+	printk(KERN_INFO "Help me Obi Pre Kenobi... Timeout in check_device_status(), deviceid %d, "
 			"desired state %d, current state %d\n",
 			deviceid, control, curr_state);
 
@@ -150,7 +150,9 @@
 			/* FCLK/ICLK present and is ON
 			 * Wait for sometime for the clocks to stabilize*/
 			DPRINTK("Adding delay\n");
+			cond_resched();
 			udelay(100);
+			cond_resched();
 			return PRCM_PASS;
 		}
 	}
@@ -264,8 +266,9 @@
 	enbit = DEV_BIT_POS(deviceid);
 	domain = DOMAIN_ID(deviceid);
 
-	if (cpu_is_omap3430() && !(omap & (AT_3430|AT_3430_ES2)))
+	if (cpu_is_omap3430() && !(omap & (AT_3430|AT_3430_ES2))) {
 		return PRCM_FAIL;
+	}
 
 	addr  = get_addr(domain, REG_AUTOIDLE);
 	valid = get_val_bits(domain, REG_AUTOIDLE);
@@ -352,9 +355,10 @@
 			return PRCM_PASS;
 
 		udelay(1);
+		cond_resched();
 	} while (retries--);
 
-	printk(KERN_INFO "Timeout in %s(), "
+	printk(KERN_INFO "We seem to be made to suffer. It's our lot in life. PCCDS: Timeout in %s(), "
 			"domainid %d, desired state %d, current state %d\n",
 			__FUNCTION__, domainid, desired_state, curr_state);
 	return PRCM_FAIL;
