Index: linux-2.6.23/fs/namei.c
===================================================================
RCS file: /ext1/sysadm/transparent/repository/linux-2.6.23/fs/namei.c,v
retrieving revision 1.2
retrieving revision 1.4
diff -u -p -r1.2 -r1.4
--- linux-2.6.23/fs/namei.c	29 Oct 2007 07:01:03 -0000	1.2
+++ linux-2.6.23/fs/namei.c	29 Oct 2007 07:05:53 -0000	1.4
@@ -2827,3 +2827,4 @@ EXPORT_SYMBOL(vfs_symlink);
 EXPORT_SYMBOL(vfs_unlink);
 EXPORT_SYMBOL(dentry_unhash);
 EXPORT_SYMBOL(generic_readlink);
+EXPORT_SYMBOL(deny_write_access);
