Index: linux-2.6.16/fs/buffer.c
===================================================================
RCS file: /proj/linux-2.6/repository/linux-2.6.16/fs/buffer.c,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -r1.1 -r1.2
--- linux-2.6.16/fs/buffer.c	20 Feb 2008 05:12:07 -0000	1.1
+++ linux-2.6.16/fs/buffer.c	20 Feb 2008 05:15:22 -0000	1.2
@@ -180,6 +180,7 @@ int fsync_super(struct super_block *sb)
 
 	return sync_blockdev(sb->s_bdev);
 }
+EXPORT_SYMBOL(fsync_super);
 
 /*
  * Write out and wait upon all dirty data associated with this
