--- linux-2.6.24-palm/drivers/cpufreq/Kconfig	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/drivers/cpufreq/Kconfig	2010-12-11 11:03:04.000000000 -0500
@@ -53,6 +53,23 @@
 
 	  If in doubt, say N.
 
+config CPU_FREQ_OVERRIDE
+	bool "Extra on-demand CPU tweaking options"
+	default y
+	help
+	  This will give options to tweak CPU settings in-demand.
+
+	  If in doubt, say Y.
+
+config CPU_FREQ_OVERRIDE_STRIPOPP
+	bool "Strip OPP1 and OPP2 from available frequencies list"
+	depends on CPU_FREQ_OVERRIDE
+	default y
+	help
+	  This will hide 125MHz and 250MHz from scaling_available_frequencies.
+
+	  If in doubt, say N.
+
 choice
 	prompt "Default CPUFreq governor"
 	default CPU_FREQ_DEFAULT_GOV_USERSPACE if CPU_FREQ_SA1100 || CPU_FREQ_SA1110
@@ -87,6 +104,13 @@
 	  program shall be able to set the CPU dynamically without having
 	  to enable the userspace governor manually.
 
+config CPU_FREQ_DEFAULT_GOV_VDEMAND
+	bool "vdemand"
+	select CPU_FREQ_GOV_VDEMAND
+	help
+	  Use the CPUFreq governor 'vdemand' as default. This will
+	  scale the CPU up and down depending on load.
+
 config CPU_FREQ_DEFAULT_GOV_ONDEMAND
 	bool "ondemand"
 	select CPU_FREQ_GOV_ONDEMAND
@@ -149,6 +173,17 @@
 
 	  If in doubt, say Y.
 
+config CPU_FREQ_GOV_VDEMAND
+	tristate "'vdemand' governor for frequency scaling"
+	help
+	  Enable this cpufreq governor to scale CPU voltage.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called cpufreq_vdemand.
+
+	  If in doubt, say Y.
+
+
 config CPU_FREQ_GOV_ONDEMAND
 	tristate "'ondemand' cpufreq policy governor"
 	select CPU_FREQ_TABLE
--- linux-2.6.24-palm/drivers/cpufreq/Makefile	2008-01-24 17:58:37.000000000 -0500
+++ linux-2.6.24-AV8B/drivers/cpufreq/Makefile	2010-12-11 11:03:40.000000000 -0500
@@ -9,7 +9,11 @@
 obj-$(CONFIG_CPU_FREQ_GOV_USERSPACE)	+= cpufreq_userspace.o
 obj-$(CONFIG_CPU_FREQ_GOV_ONDEMAND)	+= cpufreq_ondemand.o
 obj-$(CONFIG_CPU_FREQ_GOV_CONSERVATIVE)	+= cpufreq_conservative.o
+obj-$(CONFIG_CPU_FREQ_GOV_VDEMAND)	+= cpufreq_vdemand.o
 
 # CPUfreq cross-arch helpers
 obj-$(CONFIG_CPU_FREQ_TABLE)		+= freq_table.o
 
+# CPUfreq override
+obj-$(CONFIG_CPU_FREQ_OVERRIDE)         += cpufreq_override.o
+
--- linux-2.6.24-palm/drivers/cpufreq/cpufreq.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/drivers/cpufreq/cpufreq.c	2011-01-15 09:54:59.000000000 -0500
@@ -32,6 +32,11 @@
 #define dprintk(msg...) cpufreq_debug_printk(CPUFREQ_DEBUG_CORE, \
 						"cpufreq-core", msg)
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+int cpufreq_override_driver_init(void);
+void cpufreq_override_driver_exit(void);
+#endif
+
 /**
  * The "cpufreq driver" - the arch- or hardware-dependent low
  * level driver of CPUFreq support, and its spinlock. This lock
@@ -453,8 +458,8 @@
 
 show_one(cpuinfo_min_freq, cpuinfo.min_freq);
 show_one(cpuinfo_max_freq, cpuinfo.max_freq);
-show_one(scaling_min_freq, min);
-show_one(scaling_max_freq, max);
+//show_one(scaling_min_freq, min);
+//show_one(scaling_max_freq, max);
 show_one(scaling_cur_freq, cur);
 
 static int __cpufreq_set_policy(struct cpufreq_policy *data,
@@ -484,8 +489,8 @@
 	return ret ? ret : count;					\
 }
 
-store_one(scaling_min_freq,min);
-store_one(scaling_max_freq,max);
+//store_one(scaling_min_freq,min);
+//store_one(scaling_max_freq,max);
 
 /**
  * show_cpuinfo_cur_freq - current CPU frequency as detected by hardware
@@ -621,15 +626,15 @@
 define_one_ro(scaling_driver);
 define_one_ro(scaling_cur_freq);
 define_one_ro(affected_cpus);
-define_one_rw(scaling_min_freq);
-define_one_rw(scaling_max_freq);
+//define_one_rw(scaling_min_freq);
+//define_one_rw(scaling_max_freq);
 define_one_rw(scaling_governor);
 
 static struct attribute * default_attrs[] = {
 	&cpuinfo_min_freq.attr,
 	&cpuinfo_max_freq.attr,
-	&scaling_min_freq.attr,
-	&scaling_max_freq.attr,
+//	&scaling_min_freq.attr,
+//	&scaling_max_freq.attr,
 	&affected_cpus.attr,
 	&scaling_governor.attr,
 	&scaling_driver.attr,
@@ -1682,6 +1687,18 @@
 	return ret;
 }
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+int cpufreq_set_policy(struct cpufreq_policy *policy)
+{
+	struct cpufreq_policy *data = cpufreq_cpu_get(0);
+	__cpufreq_set_policy(data,policy);
+	data->user_policy.min = data->min;
+	data->user_policy.max = data->max;
+	cpufreq_cpu_put(data);
+}
+EXPORT_SYMBOL(cpufreq_set_policy);
+#endif
+
 /**
  *	cpufreq_update_policy - re-evaluate an existing cpufreq policy
  *	@cpu: CPU which shall be re-evaluated
@@ -1831,6 +1848,10 @@
 		cpufreq_debug_enable_ratelimit();
 	}
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+	cpufreq_override_driver_init();
+#endif
+
 	return (ret);
 }
 EXPORT_SYMBOL_GPL(cpufreq_register_driver);
@@ -1864,6 +1885,10 @@
 	cpufreq_driver = NULL;
 	spin_unlock_irqrestore(&cpufreq_driver_lock, flags);
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+	cpufreq_override_driver_exit();
+#endif
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(cpufreq_unregister_driver);
--- linux-2.6.24-palm/include/linux/cpufreq.h	2010-06-11 17:48:35.000000000 -0400
+++ linux-2.6.24-AV8B/include/linux/cpufreq.h	2010-12-11 11:08:53.000000000 -0500
@@ -309,6 +309,9 @@
 #elif defined(CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE)
 extern struct cpufreq_governor cpufreq_gov_userspace;
 #define CPUFREQ_DEFAULT_GOVERNOR	(&cpufreq_gov_userspace)
+#elif defined(CONFIG_CPU_FREQ_DEFAULT_GOV_VDEMAND)
+extern struct cpufreq_governor cpufreq_gov_vdemand;
+#define CPUFREQ_DEFAULT_GOVERNOR	(&cpufreq_gov_vdemand)
 #elif defined(CONFIG_CPU_FREQ_DEFAULT_GOV_ONDEMAND)
 extern struct cpufreq_governor cpufreq_gov_ondemand;
 #define CPUFREQ_DEFAULT_GOVERNOR	(&cpufreq_gov_ondemand)
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/prcm_opp.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/prcm_opp.c	2011-02-19 10:07:43.000000000 -0500
@@ -57,12 +57,12 @@
 static u8 mpu_iva2_vdd1_volts [PRCM_NO_VDD1_OPPS] = {
 	/* Vsel corresponding to 0.975V (OPP1), 1.050V (OPP2),
 				1.20V (OPP3), 1.27V (OPP4), 1.35 (OPP5) */
-	0x1e, 0x24, 0x30, 0x36, 0x3C
+	0x1e, 0x24, 0x30, 0x36, 0x43
 };
 
 static u8 core_l3_vdd2_volts [PRCM_NO_VDD2_OPPS] = { /* only 3 OPPs */
 	/* Vsel corresponding to 0.975V (OPP1), 1.05V (OPP2), 1.15 (OPP3) */
-	0x1e, 0x24, 0x2c
+	0x1e, 0x24, 0x2a
 };
 
 
@@ -161,8 +161,8 @@
 	{{0x1F4, 0x0C, 0x03, 0x04}, {0x1F4, 0x0C, 0x03, 0x02},
 	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
 	{0x1F4, 0x0C, 0x03, 0x01}, {0x226, 0x0C, 0x03, 0x01},
-	/* OPP5 (600 Mhz) */
-	{0x258, 0x0C, 0x03, 0x01} },
+	/* OPP5 (1005 Mhz) */
+	{0x3ED, 0x0C, 0x03, 0x01} },
 	/* 19.2M values */
 	/* OPP1(125 Mhz) and OPP2(250 Mhz)*/
 	{{0x271, 0x17, 0x03, 0x04}, {0x271, 0x17, 0x03, 0x02},
@@ -476,7 +476,7 @@
 			valid_rate = clk_round_rate(p_vdd1_clk, S550M);
 			break;
 		case PRCM_VDD1_OPP5:
-			valid_rate = clk_round_rate(p_vdd1_clk, S600M);
+			valid_rate = clk_round_rate(p_vdd1_clk, S1005M);
 			break;
 		}
 		p_vdd1_clk->set_rate(p_vdd1_clk, valid_rate);
@@ -494,7 +494,7 @@
 	{250, 180, CO_VDD1_OPP2, PRCM_VDD1_OPP2},
 	{500, 360, CO_VDD1_OPP3, PRCM_VDD1_OPP3},
 	{550, 396, CO_VDD1_OPP4, PRCM_VDD1_OPP4},
-	{600, 430, CO_VDD1_OPP5, PRCM_VDD1_OPP5},
+	{1005, 520, CO_VDD1_OPP5, PRCM_VDD1_OPP5},
 };
 static struct vdd2_core_freq_d {
 	unsigned int freq;
@@ -507,7 +507,7 @@
 };
 
 static unsigned int rnd_rate_vdd1[5] = {
-	S125M, S250M, S500M, S550M, S600M
+	S125M, S250M, S500M, S550M, S1005M
 };
 static unsigned int rnd_rate_vdd2[3] = {
 	0, S83M, S166M
@@ -1523,3 +1523,40 @@
 
 	return -1;
 }
+
+void omap_pm_opp_get_volts(u8 *vdd1_volts[]) {
+       memcpy(vdd1_volts,mpu_iva2_vdd1_volts,sizeof(mpu_iva2_vdd1_volts));
+}
+EXPORT_SYMBOL(omap_pm_opp_get_volts);
+
+void omap_pm_opp_set_volts(u8 vdd1_volts[]) {
+       memcpy(mpu_iva2_vdd1_volts,vdd1_volts,sizeof(mpu_iva2_vdd1_volts));
+       prcm_do_voltage_scaling(current_vdd1_opp, current_vdd1_opp-1);
+}
+EXPORT_SYMBOL(omap_pm_opp_set_volts);
+
+void omap_pm_opp_get_vdd2_volts(u8 *vdd2_volts[]) {
+        memcpy(vdd2_volts,core_l3_vdd2_volts,sizeof(core_l3_vdd2_volts));
+}
+EXPORT_SYMBOL(omap_pm_opp_get_vdd2_volts);
+
+void omap_pm_opp_set_vdd2_volts(u8 vdd2_volts[]) {
+        memcpy(core_l3_vdd2_volts,vdd2_volts,sizeof(core_l3_vdd2_volts));
+        prcm_do_voltage_scaling(current_vdd2_opp, current_vdd2_opp-1);
+}
+EXPORT_SYMBOL(omap_pm_opp_set_vdd2_volts);
+
+void omap_pm_opp_get_vdd2_freq(u8 *vdd2_freqs[]) {
+       int i;
+       u8 f[MAX_VDD2_OPP];
+
+       for(i=0;i < MAX_VDD2_OPP;i++)
+               f[i]=(u8 )vdd2_core_freq[i].freq;
+       memcpy(vdd2_freqs,f,sizeof(f));
+}
+EXPORT_SYMBOL(omap_pm_opp_get_vdd2_freq);
+
+unsigned int prcm_get_current_vdd1_opp_no(void) {
+	return get_opp_no(current_vdd1_opp);
+}
+EXPORT_SYMBOL(prcm_get_current_vdd1_opp_no);
--- linux-2.6.24-palm/drivers/cpufreq/cpufreq_override.c	1969-12-31 19:00:00.000000000 -0500
+++ linux-2.6.24-AV8B/drivers/cpufreq/cpufreq_override.c	2011-02-19 18:40:32.000000000 -0500
@@ -0,0 +1,345 @@
+/*
+ *  drivers/cpufreq/cpufreq_override.c
+ *
+ *  	Marco Benton <marco@unixpsycho.com>.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include <linux/kernel.h>
+#include <linux/sysdev.h>
+#include <linux/cpu.h>
+#include <linux/sysfs.h>
+#include <linux/cpufreq.h>
+#include <linux/jiffies.h>
+#include <linux/kobject.h>
+#include <linux/workqueue.h>
+
+// VDD1 Vsel max
+#define VDD1_VSEL_MAX 112
+
+// VDD1 Vsel min
+#define VDD1_VSEL_MIN 25
+
+// VDD2 Vsel max
+#define VDD2_VSEL_MAX 55
+
+// VDD2 Vsel min
+#define VDD2_VSEL_MIN 25
+
+// High temp alarm and cap
+#define HIGHTEMP_SCALEBACK 55
+
+//Reset temp from alarm
+#define LOWTEMP_RESET 50
+
+// Polling frequency jiffies
+#define BATTERY_POLLING 30000
+#define TEMP_POLLING 1000
+#define CHARGER_POLLING 3000
+
+// Battery scaleback percent
+#define BATTERY_PERCENT 25
+
+// Battery scaleback speed
+#define SCALEBACK_SPEED 500000
+
+void omap_pm_opp_get_volts(u8 *vdd1_volts[]);
+void omap_pm_opp_set_volts(u8 vdd1_volts[]);
+void omap_pm_opp_get_vdd2_volts(u8 *vdd2_volts[]);
+void omap_pm_opp_set_vdd2_volts(u8 vdd2_volts[]);
+void omap_pm_opp_get_vdd2_freq(u8 *vdd2_freqs[]);
+int omap34xx_get_temp(void);
+int cpufreq_set_policy(struct cpufreq_policy *policy);
+int ds2784_getpercent(int *ret_percent);
+int gadget_event_state_current(void);
+int ds2784_getcurrent(int *ret_current);
+
+static inline void check_temp(struct work_struct *work);
+static inline void check_battery(struct work_struct *work);
+static inline void check_charger(struct work_struct *work);
+
+static u32 override_hightemp=HIGHTEMP_SCALEBACK;
+static u32 override_lowtemp=LOWTEMP_RESET;
+static int battery_scaleback_percent=BATTERY_PERCENT;
+
+u8 overtemp_alarm=0;
+u8 battery_alarm=0;
+u8 charging_state=0;
+u8 ch_override=0;
+
+static DEFINE_MUTEX(override_mutex);
+static DECLARE_DELAYED_WORK(dbs_work, check_temp);
+static DECLARE_DELAYED_WORK(dbs_work2, check_battery);
+static DECLARE_DELAYED_WORK(dbs_work3, check_charger);
+
+#define CPUFREQ_OVERRIDE_ATTR(_name,_mode,_show,_store) \
+static struct freq_attr _attr_##_name = {\
+        .attr = {.name = __stringify(_name), .mode = _mode, }, \
+        .show = _show,\
+        .store = _store,\
+};
+
+#define CPUFREQ_OVERRIDE_ATTR2(_name,_mode,_show) \
+static struct freq_attr _attr_##_name = {\
+	.attr = {.name = __stringify(_name), .mode = _mode, }, \
+	.show = _show,\
+};
+
+static void change_freq_low(void) {
+	struct cpufreq_policy new_policy;
+
+	cpufreq_get_policy(&new_policy,0);
+	new_policy.min=SCALEBACK_SPEED;
+	new_policy.max=SCALEBACK_SPEED;
+	cpufreq_set_policy(&new_policy);
+}
+
+static void change_freq_high(void) {
+	struct cpufreq_policy new_policy;
+
+	cpufreq_get_policy(&new_policy,0);
+	new_policy.min=1005000;
+	new_policy.max=1005000;
+	cpufreq_set_policy(&new_policy);
+}
+
+static inline void check_charger(struct work_struct *work) {
+	int cur, current_mA;
+
+	if(overtemp_alarm || battery_alarm) goto n1;
+
+	ds2784_getcurrent(&cur);
+	current_mA=gadget_event_state_current();
+
+	if((cur>0) && (current_mA < 500)) {
+		// Assume Touchstone
+		if(!charging_state) {
+			charging_state=1;
+			change_freq_low();
+			printk("CPUfreq: TS found!\n");
+		}
+	} 
+	else {
+		if(current_mA == 1000) {
+			if(!charging_state) {
+				charging_state=1;
+				change_freq_low();
+				printk("CPUfreq: 1000mA charger found!\n");
+			}
+		}
+		else {
+			if(charging_state) {
+				charging_state=0;
+				change_freq_high();
+				printk("CPUfreq: charger unplugged!\n");
+			}
+		}
+	}
+n1:
+	 schedule_delayed_work(&dbs_work3,CHARGER_POLLING);
+}
+
+static inline void check_temp(struct work_struct *work) {
+        u32 cputemp;
+
+	if(charging_state || battery_alarm) goto out;
+
+        cputemp = omap34xx_get_temp();    // Get CPU temp
+
+	// Check values in case driver hasnt polled
+	cputemp = (cputemp < 100) ? cputemp : 0;
+
+        if(cputemp > override_hightemp) {
+                if(!overtemp_alarm) {
+                        printk("CPUfreq: CPU temp warning! %dC\n",cputemp);
+                        overtemp_alarm = 1;
+			change_freq_low();
+                }
+        }
+        else {
+                if((overtemp_alarm) && (cputemp < override_lowtemp)) {
+                        printk("CPUfreq: CPU temp back under control! %dC\n",
+								cputemp);
+                        if (overtemp_alarm) {
+				change_freq_high();
+                        	overtemp_alarm = 0;
+			}
+                }
+        }
+out:
+	schedule_delayed_work(&dbs_work,TEMP_POLLING);
+}
+
+static inline void check_battery(struct work_struct *work) { 
+        int battery_per;
+
+        if(charging_state || overtemp_alarm) goto out;
+
+        ds2784_getpercent(&battery_per);  // Get battery percent left
+
+	if(battery_per<battery_scaleback_percent) {
+		if(!battery_alarm) {
+			printk("CPUFreq: battery low! < %d%%\n",battery_per);
+			battery_alarm = 1;
+			change_freq_low();
+		}
+	}
+	else {
+		if(battery_alarm) {
+			printk("CPUFreq: battery OK\n");
+			change_freq_high();
+			battery_alarm = 0;
+		}
+	}
+
+out:
+	schedule_delayed_work(&dbs_work2,BATTERY_POLLING);
+}
+
+static ssize_t show_vdd1_vsel_max(struct cpufreq_policy *policy, char *buf) {
+	return sprintf(buf, "%hu\n",VDD1_VSEL_MAX);
+}
+
+static ssize_t show_vdd1_vsel_min(struct cpufreq_policy *policy, char *buf) {
+	return sprintf(buf, "%hu\n",VDD1_VSEL_MIN);
+}
+CPUFREQ_OVERRIDE_ATTR2(vdd1_vsel_max,0444,show_vdd1_vsel_max);
+CPUFREQ_OVERRIDE_ATTR2(vdd1_vsel_min,0444,show_vdd1_vsel_min);
+
+static ssize_t show_vdd2_vsel_max(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%hu\n",VDD2_VSEL_MAX);
+}
+
+static ssize_t show_vdd2_vsel_min(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%hu\n",VDD2_VSEL_MIN);
+}
+CPUFREQ_OVERRIDE_ATTR2(vdd2_vsel_min,0444,show_vdd2_vsel_min);
+CPUFREQ_OVERRIDE_ATTR2(vdd2_vsel_max,0444,show_vdd2_vsel_max);
+
+static ssize_t show_ch_override(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%hu\n",ch_override);
+}
+
+static ssize_t store_ch_override(struct cpufreq_policy *policy,
+						char *buf, size_t count) {
+	u8 override=0;
+
+	if(sscanf(buf, "%hhu", &override) == 1) {
+		override=(override) ? 1 : 0;
+		if(override) {
+			charging_state=0;
+			if(!ch_override) cancel_delayed_work(&dbs_work3);
+			change_freq_high();
+			printk("CPUfreq: charger override enabled\n");
+		}
+		else
+			if(ch_override)
+				schedule_delayed_work(&dbs_work3,CHARGER_POLLING);
+		ch_override=override;
+	}
+	else printk("CPUfreq: invalid charger override value\n");
+
+	return count;
+}
+
+static ssize_t show_hightemp_scaleback(struct cpufreq_policy *policy,
+						char *buf) {
+        return sprintf(buf, "%d\n", override_hightemp);
+}
+
+static ssize_t store_hightemp_scaleback(struct cpufreq_policy *policy,
+						char *buf, size_t count) {
+        int maxtemp=0;
+
+        if(sscanf(buf, "%d", &maxtemp) == 1)
+                override_hightemp=(maxtemp) ? maxtemp : HIGHTEMP_SCALEBACK;
+        else printk("CPUfreq: invalid max temp\n");
+
+	return count;
+}
+
+static ssize_t show_battery_scaleback_per(struct cpufreq_policy *policy,
+                                                char *buf) {
+	return sprintf(buf, "%d\n", battery_scaleback_percent);
+}
+
+static ssize_t store_battery_scaleback_per(struct cpufreq_policy *policy,
+                                                char *buf, size_t count) {
+	int bat=0;
+
+	if(sscanf(buf, "%d", &bat) == 1)
+		battery_scaleback_percent=((bat > -1) && (bat<100))
+						? bat : BATTERY_PERCENT;
+	else printk("CPUfreq: invalid battery percentage\n");
+
+	return count;
+}
+
+static ssize_t show_lowtemp_reset(struct cpufreq_policy *policy, char *buf) {
+
+        return sprintf(buf, "%d\n", override_lowtemp);
+}
+
+static ssize_t store_lowtemp_reset(struct cpufreq_policy *policy, char *buf,
+						size_t count) {
+        int lowtemp=0;
+
+        if(sscanf(buf, "%d", &lowtemp) == 1)
+                override_lowtemp=(lowtemp) ? lowtemp : LOWTEMP_RESET;
+        else printk("CPUfreq: invalid low temp\n");
+
+	return count;
+}
+
+CPUFREQ_OVERRIDE_ATTR(battery_scaleback_percent,0644,
+			show_battery_scaleback_per,
+			store_battery_scaleback_per);
+CPUFREQ_OVERRIDE_ATTR(cpu_hightemp_alarm,0644,show_hightemp_scaleback,
+			store_hightemp_scaleback);
+CPUFREQ_OVERRIDE_ATTR(cpu_hightemp_reset,0644,show_lowtemp_reset,
+			store_lowtemp_reset);
+CPUFREQ_OVERRIDE_ATTR(override_charger,0644,show_ch_override,
+			store_ch_override);
+
+static struct attribute *default_attrs[] = {
+        &_attr_vdd1_vsel_min.attr,
+        &_attr_vdd1_vsel_max.attr,
+	&_attr_vdd2_vsel_min.attr,
+	&_attr_vdd2_vsel_max.attr,
+        &_attr_cpu_hightemp_alarm.attr,
+        &_attr_cpu_hightemp_reset.attr,
+	&_attr_battery_scaleback_percent.attr,
+	&_attr_override_charger.attr,
+        NULL
+};
+
+static struct attribute_group override_attr_group = {
+        .attrs = default_attrs,
+        .name = "override"
+};
+
+int cpufreq_override_driver_init(void) {
+        struct cpufreq_policy *data = cpufreq_cpu_get(0);
+	schedule_delayed_work(&dbs_work,TEMP_POLLING);
+	schedule_delayed_work(&dbs_work2,BATTERY_POLLING);
+	schedule_delayed_work(&dbs_work3,CHARGER_POLLING);
+        return sysfs_create_group(&data->kobj,&override_attr_group);
+}
+EXPORT_SYMBOL(cpufreq_override_driver_init);
+
+void cpufreq_override_driver_exit(void) {
+	struct cpufreq_policy *policy = cpufreq_cpu_get(0);
+	cancel_delayed_work(&dbs_work);
+	cancel_delayed_work(&dbs_work2);
+	cancel_delayed_work(&dbs_work3);
+	sysfs_remove_group(&policy->kobj, &override_attr_group);
+	flush_scheduled_work();
+}
+EXPORT_SYMBOL(cpufreq_override_driver_exit);
+
+MODULE_AUTHOR ("marco@unixpsycho.com");
+MODULE_DESCRIPTION ("'cpufreq_override' - A driver to do cool stuff ");
+MODULE_LICENSE ("GPL");
--- linux-2.6.24-palm/drivers/cpufreq/cpufreq_vdemand.c	1969-12-31 19:00:00.000000000 -0500
+++ linux-2.6.24-AV8B/drivers/cpufreq/cpufreq_vdemand.c	2011-03-05 08:56:03.000000000 -0500
@@ -0,0 +1,351 @@
+/*
+ *  linux/drivers/cpufreq/cpufreq_vdemand.c
+ *
+ *  Marco Benton marco@unixpsycho.com 
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/spinlock.h>
+#include <linux/interrupt.h>
+#include <linux/cpufreq.h>
+#include <linux/cpu.h>
+#include <linux/types.h>
+#include <linux/mutex.h>
+#include <asm/uaccess.h>
+#include <linux/workqueue.h>
+#include <linux/kernel_stat.h>
+#include <asm/arch/prcm.h>
+
+// VDD1 Vsel max
+#define VDD1_VSEL_MAX 112
+
+// VDD1 Vsel min
+#define VDD1_VSEL_MIN 25
+
+// VDD2 Vsel max
+#define VDD2_VSEL_MAX 55
+
+// VDD2 Vsel min
+#define VDD2_VSEL_MIN 33
+
+// VDD defaults
+#define VDD1_VSEL { 0x40, 0x42, 0x45 }
+#define VDD2_VSEL { 0x26, 0x28, 0x2b }
+
+// vdemand poll rate
+#define VDEMAND_POLL_RATE 150
+
+// vdemand poll rate min
+#define VDEMAND_POLL_RATE_MIN 10
+
+// vdemand poll rate max
+#define VDEMAND_POLL_RATE_MAX 1000
+
+void omap_pm_opp_set_volts(u8 vdd1_volts[]);
+void omap_pm_opp_get_volts(u8 *vdd1_volts[]);
+void omap_pm_opp_set_vdd2_volts(u8 vdd2_volts[]);
+void omap_pm_opp_get_vdd2_volts(u8 *vdd2_volts[]);
+int prcm_get_current_vdd1_opp_no(void);
+
+int cpufreq_set_policy(struct cpufreq_policy *policy);
+
+static unsigned int cpu_is_managed=0,last_load=0;
+static cputime64_t prev_cpu_wall=0,prev_cpu_idle=0;
+static u8 v_volt[3] = VDD1_VSEL;
+static u8 v_volt_vdd2[3] = VDD2_VSEL;
+static u8 vdd1_vsel_def[3] = VDD1_VSEL;
+static u8 vdd2_vsel_def[3] = VDD2_VSEL;
+static unsigned int vdemand_poll_rate=VDEMAND_POLL_RATE;
+
+static inline void check_load(struct work_struct *work);
+
+static DEFINE_MUTEX(vdemand_mutex);
+
+static DECLARE_DELAYED_WORK(dbs_work, check_load);
+
+#define CPUFREQ_OVERRIDE_ATTR(_name,_mode,_show,_store) \
+static struct freq_attr _attr_##_name = {\
+	.attr = {.name = __stringify(_name), .mode = _mode, }, \
+	.show = _show,\
+	.store = _store,\
+};
+
+#define CPUFREQ_OVERRIDE_ATTR2(_name,_mode,_show) \
+static struct freq_attr _attr_##_name = {\
+        .attr = {.name = __stringify(_name), .mode = _mode, }, \
+        .show = _show,\
+};
+
+static ssize_t show_vdemand_vsel_vdd1_def(struct cpufreq_policy *policy,
+								char *buf) {
+        return sprintf(buf, "%hu %hu %hu\n", vdd1_vsel_def[2],vdd1_vsel_def[1],
+						vdd1_vsel_def[0]);
+}
+
+static ssize_t show_vdemand_vsel_vdd2_def(struct cpufreq_policy *policy,
+								char *buf) {
+        return sprintf(buf, "%hu %hu %hu\n", vdd2_vsel_def[2],vdd2_vsel_def[1],
+						vdd2_vsel_def[0]);
+}
+CPUFREQ_OVERRIDE_ATTR2(vdemand_vsel_vdd1_default,0444,
+			show_vdemand_vsel_vdd1_def);
+CPUFREQ_OVERRIDE_ATTR2(vdemand_vsel_vdd2_default,0444,
+			show_vdemand_vsel_vdd2_def);
+
+static ssize_t show_vdemand_vsel(struct cpufreq_policy *policy, char *buf) {
+	return sprintf(buf, "%hu %hu %hu\n", v_volt[2],v_volt[1],v_volt[0]);
+}
+
+static ssize_t store_vdemand_vsel(struct cpufreq_policy *policy, char *buf,
+                                                size_t count) {
+        u8 volt[3], i;
+
+        mutex_lock(&vdemand_mutex);
+
+        if(sscanf(buf, "%hhu %hhu %hhu", &volt[2],&volt[1],&volt[0])) {
+                for(i=0;i < 3;i++) {
+                        if((volt[i] < VDD1_VSEL_MIN) ||
+				(volt[i] > VDD1_VSEL_MAX)) {
+                                printk("CPUfreq: invalid vsel! Resetting to default!\n");
+				volt[i]=vdd1_vsel_def[i];
+                                break;
+                        }
+                }
+                if(i == 3) memcpy(v_volt,volt,sizeof(volt));
+        }
+        else printk("CPUfreq: missing vsel values\n");
+
+        mutex_unlock(&vdemand_mutex);
+        return count;
+}
+
+static ssize_t show_vdemand_poll(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%hu\n", vdemand_poll_rate);
+}
+
+static ssize_t store_vdemand_poll(struct cpufreq_policy *policy, char *buf,
+                                                size_t count) {
+        unsigned int i;
+
+        if(sscanf(buf, "%u", &i)) {
+                if((i < VDEMAND_POLL_RATE_MIN) && (i > VDEMAND_POLL_RATE_MAX))
+                        printk("CPUfreq: invalid \n");
+                else {
+                        vdemand_poll_rate=i;
+                }
+        }
+        else printk("CPUfreq: missing value\n");
+
+        return count;
+}
+
+CPUFREQ_OVERRIDE_ATTR(vdemand_vsel_vdd1,0644,show_vdemand_vsel,
+						store_vdemand_vsel);
+CPUFREQ_OVERRIDE_ATTR(vdemand_poll_rate,0644,show_vdemand_poll,
+						store_vdemand_poll);
+
+static ssize_t show_vdemand_vsel_vdd2(struct cpufreq_policy *policy,
+								char *buf) {
+        return sprintf(buf, "%hu %hu %hu\n", v_volt_vdd2[2],v_volt_vdd2[1],
+							v_volt_vdd2[0]);
+}
+
+static ssize_t show_vdemand_poll_rate_min(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%u\n",VDEMAND_POLL_RATE_MIN);
+}
+
+static ssize_t show_vdemand_poll_rate_max(struct cpufreq_policy *policy, char *buf) {
+        return sprintf(buf, "%u\n", VDEMAND_POLL_RATE_MAX);
+}
+CPUFREQ_OVERRIDE_ATTR2(vdemand_poll_rate_min,0444,show_vdemand_poll_rate_min);
+CPUFREQ_OVERRIDE_ATTR2(vdemand_poll_rate_max,0444,show_vdemand_poll_rate_max);
+
+static ssize_t store_vdemand_vsel_vdd2(struct cpufreq_policy *policy, char *buf,
+                                                size_t count) {
+        u8 volt[3], i;
+
+        mutex_lock(&vdemand_mutex);
+
+        if(sscanf(buf, "%hhu %hhu %hhu", &volt[2],&volt[1],&volt[0])) {
+                for(i=0;i < 3;i++) {
+                        if((volt[i] < VDD2_VSEL_MIN) ||
+                                (volt[i] > VDD2_VSEL_MAX)) {
+                                printk("CPUfreq: invalid vsel! Resetting to default!\n");
+				volt[i]=vdd2_vsel_def[i];
+                                break;
+                        }
+                }
+                if(i == 3) memcpy(v_volt_vdd2,volt,sizeof(volt));
+        }
+        else printk("CPUfreq: missing vsel values\n");
+
+        mutex_unlock(&vdemand_mutex);
+        return count;
+}
+
+CPUFREQ_OVERRIDE_ATTR(vdemand_vsel_vdd2,0644,show_vdemand_vsel_vdd2,
+						store_vdemand_vsel_vdd2);
+
+static struct attribute *default_attrs[] = {
+        &_attr_vdemand_vsel_vdd1.attr,
+        &_attr_vdemand_vsel_vdd2.attr,
+        &_attr_vdemand_poll_rate.attr,
+        &_attr_vdemand_vsel_vdd1_default.attr,
+        &_attr_vdemand_vsel_vdd2_default.attr,
+	&_attr_vdemand_poll_rate_min.attr,
+	&_attr_vdemand_poll_rate_max.attr,
+        NULL
+};
+
+static struct attribute_group vdemand_attr_group = {
+        .attrs = default_attrs,
+        .name = "vdemand"
+};
+
+static inline cputime64_t get_cpu_idle_time(unsigned int cpu) {
+        cputime64_t idle_time;
+        cputime64_t cur_jiffies;
+        cputime64_t busy_time;
+
+        cur_jiffies = jiffies64_to_cputime64(get_jiffies_64());
+        busy_time = cputime64_add(kstat_cpu(cpu).cpustat.user,
+                        kstat_cpu(cpu).cpustat.system);
+
+        busy_time = cputime64_add(busy_time,kstat_cpu(cpu).cpustat.irq);
+        busy_time = cputime64_add(busy_time,kstat_cpu(cpu).cpustat.softirq);
+        busy_time = cputime64_add(busy_time,kstat_cpu(cpu).cpustat.steal);
+
+        idle_time = cputime64_sub(cur_jiffies, busy_time);
+        return idle_time;
+}
+
+static inline void check_load(struct work_struct *work) {
+	unsigned int tmp_idle_ticks,idle_ticks,total_ticks,load=0;
+	cputime64_t total_idle_ticks,cur_jiffies;
+	u8 volt[PRCM_NO_VDD1_OPPS],volt_vdd2[3];
+
+	mutex_lock(&vdemand_mutex);
+
+	if(prcm_get_current_vdd1_opp_no() < PRCM_NO_VDD1_OPPS) goto out;
+
+	omap_pm_opp_get_volts(&volt);
+	omap_pm_opp_get_vdd2_volts(&volt_vdd2);
+
+	idle_ticks = UINT_MAX;
+        cur_jiffies = jiffies64_to_cputime64(get_jiffies_64());
+        total_ticks = (unsigned int) cputime64_sub(cur_jiffies,prev_cpu_wall);
+
+        if (!total_ticks) goto out;
+
+	prev_cpu_wall = get_jiffies_64();
+	total_idle_ticks = get_cpu_idle_time(0);
+	tmp_idle_ticks = (unsigned int )cputime64_sub(total_idle_ticks,
+		prev_cpu_idle);
+	prev_cpu_idle = total_idle_ticks;
+
+	if(tmp_idle_ticks < idle_ticks) idle_ticks = tmp_idle_ticks;
+	if(total_ticks > idle_ticks)
+		load = (100 * (total_ticks - idle_ticks)) / total_ticks;
+
+	if(!last_load) goto out;
+
+	if((load < 30) && (last_load > 30)) {
+		volt[PRCM_NO_VDD1_OPPS-1]=v_volt[0];
+		volt_vdd2[2]=v_volt_vdd2[0];
+		omap_pm_opp_set_volts(volt);
+		omap_pm_opp_set_vdd2_volts(volt_vdd2);
+	}
+        if(((load > 30) && (load < 70)) &&
+	    ((last_load < 30) || (last_load > 70))) {
+                volt[PRCM_NO_VDD1_OPPS-1]=v_volt[1];
+		volt_vdd2[2]=v_volt_vdd2[1];
+                omap_pm_opp_set_volts(volt);
+		omap_pm_opp_set_vdd2_volts(volt_vdd2);
+        }
+	if((load > 70) && (last_load < 70)) {
+                volt[PRCM_NO_VDD1_OPPS-1]=v_volt[2];
+		volt_vdd2[2]=v_volt_vdd2[2];
+                omap_pm_opp_set_volts(volt);
+		omap_pm_opp_set_vdd2_volts(volt_vdd2);
+        }
+
+out:
+	last_load=load;
+	mutex_unlock(&vdemand_mutex);
+	schedule_delayed_work(&dbs_work,vdemand_poll_rate);
+
+	return;
+}
+
+static int cpufreq_governor_vdemand(struct cpufreq_policy *policy,
+				   unsigned int event) {
+	int rc;
+	struct cpufreq_policy new_policy;
+
+	switch (event) {
+		case CPUFREQ_GOV_START:
+			if(cpu_is_managed) break;
+
+			cpu_is_managed = 1;
+			prev_cpu_idle = get_cpu_idle_time(0);
+			prev_cpu_wall = get_jiffies_64();
+
+			struct cpufreq_policy *data = cpufreq_cpu_get(0);
+			rc=sysfs_create_group(&data->kobj,&vdemand_attr_group);
+
+                        cpufreq_get_policy(&new_policy,0);
+			new_policy.min=1005000;
+			new_policy.max=1005000;
+			cpufreq_set_policy(&new_policy);
+			schedule_delayed_work(&dbs_work,vdemand_poll_rate);
+			break;
+		case CPUFREQ_GOV_STOP:
+			cpu_is_managed = 0;
+			cancel_delayed_work(&dbs_work);
+			sysfs_remove_group(&policy->kobj, &vdemand_attr_group);
+			break;
+		case CPUFREQ_GOV_LIMITS:
+			mutex_lock(&vdemand_mutex);
+			printk("vdemand: policy change\n");
+			__cpufreq_driver_target(policy,
+					policy->max, CPUFREQ_RELATION_H);
+			mutex_unlock(&vdemand_mutex);
+			break;
+		}
+		return 0;
+}
+
+struct cpufreq_governor cpufreq_gov_vdemand = {
+	.name		= "vdemand",
+	.governor	= cpufreq_governor_vdemand,
+	.owner		= THIS_MODULE,
+};
+
+static int __init cpufreq_gov_vdemand_init(void) {
+	return cpufreq_register_governor(&cpufreq_gov_vdemand);
+}
+
+static void __exit cpufreq_gov_vdemand_exit(void) {
+	flush_scheduled_work();
+	cpufreq_unregister_governor(&cpufreq_gov_vdemand);
+}
+
+EXPORT_SYMBOL(cpufreq_gov_vdemand);
+
+MODULE_AUTHOR ("marco@unixpsycho.com");
+MODULE_DESCRIPTION ("CPUfreq policy governor 'vdemand'");
+MODULE_LICENSE ("GPL");
+
+#ifdef CONFIG_CPU_FREQ_DEFAULT_GOV_VDEMAND
+fs_initcall(cpufreq_gov_vdemand_init);
+#else
+module_init(cpufreq_gov_vdemand_init);
+#endif
+module_exit(cpufreq_gov_vdemand_exit);
--- linux-2.6.24-palm/drivers/hwmon/omap34xx_temp.c     2010-05-31 12:05:29.000000000 -0400
+++ linux-2.6.24-AV8B/drivers/hwmon/omap34xx_temp.c   2010-07-11 20:46:13.000000000 -0400
@@ -140,6 +140,14 @@
	mutex_unlock(&data->update_lock);
 }

+int omap34xx_get_temp(void) {
+	struct omap34xx_data *data =
+		dev_get_drvdata(&omap34xx_temp_device.dev);
+	omap34xx_update(data);
+	return adc_to_temp[omap_ctrl_readl(OMAP343X_CONTROL_TEMP_SENSOR) & ((1<<7) - 1)];
+}
+EXPORT_SYMBOL(omap34xx_get_temp);
+
 static ssize_t show_name(struct device *dev,
 		struct device_attribute *devattr, char *buf)
 {
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/clock.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/clock.c	2010-12-11 13:59:22.000000000 -0500
@@ -709,6 +709,9 @@
 
 	prcm = vdd1_rate_table + ARRAY_SIZE(vdd1_rate_table) -1;
 	for (; prcm->speed; prcm--) {
+#ifdef CONFIG_CPU_FREQ_OVERRIDE_STRIPOPP
+		if((prcm->speed / 1000) < 500000) continue;
+#endif
 		freq_table[i].index = i;
 		freq_table[i].frequency = prcm->speed / 1000;
 		i++;
--- linux-2.6.24-palm/drivers/usb/gadget/gadget_event.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/drivers/usb/gadget/gadget_event.c	2011-01-12 00:49:49.000000000 -0500
@@ -282,6 +282,12 @@
 }
 EXPORT_SYMBOL(gadget_event_power_state_changed);
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+int gadget_event_state_current(void) {
+	return the_state.current_mA;
+}
+#endif
+
 static int __init init(void)
 {
 	int ret = 0;
--- linux-2.6.24-palm/drivers/w1/slaves/w1_ds2784.c	2010-06-11 17:48:35.000000000 -0400
+++ linux-2.6.24-AV8B/drivers/w1/slaves/w1_ds2784.c	2011-01-12 00:50:50.000000000 -0500
@@ -1133,11 +1133,20 @@
 */
 static struct device *battery_device = NULL;
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+int ds2784_getpercent(int *ret_percent)
+{
+        if (!battery_device) return -1;
+        return ds2784_getpercent_dev(battery_device, ret_percent);
+}
+EXPORT_SYMBOL(ds2784_getpercent);
+#else
 static int ds2784_getpercent(int *ret_percent)
 {
 	if (!battery_device) return -1;
 	return ds2784_getpercent_dev(battery_device, ret_percent);
 }
+#endif
 
 static int ds2784_getvoltage(int *ret_voltage)
 {
@@ -1151,11 +1160,20 @@
 	return ds2784_gettemperature_dev(battery_device, ret_temperature);
 }
 
+#ifdef CONFIG_CPU_FREQ_OVERRIDE
+int ds2784_getcurrent(int *ret_current)
+{
+        if (!battery_device) return -1;
+        return ds2784_getcurrent_dev(battery_device, ret_current);
+}
+EXPORT_SYMBOL(ds2784_getcurrent);
+#else
 static int ds2784_getcurrent(int *ret_current)
 {
 	if (!battery_device) return -1;
 	return ds2784_getcurrent_dev(battery_device, ret_current);
 }
+#endif
 
 static struct battery_ops ds2784_battery_ops = {
 	.get_percent       = ds2784_getpercent,
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/clock.h	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/clock.h	2010-12-11 11:33:24.000000000 -0500
@@ -48,8 +48,7 @@
 #define S250M	250000000
 #define S500M	500000000
 #define S550M	550000000
-#define S625M	625000000
-#define S600M	600000000
+#define S1005M	1005000000
 
 /* Macro to enable clock control via clock framework */
 #define ENABLE_CLOCKCONTROL 1
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/clock_tree.h	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/clock_tree.h	2010-12-11 11:32:38.000000000 -0500
@@ -92,7 +92,7 @@
 	/*OPP4*/
 	{S550M, PRCM_VDD1_OPP4, RATE_IN_343X},
 	/*OPP5*/
-	{S600M, PRCM_VDD1_OPP5, RATE_IN_343X},
+	{S1005M, PRCM_VDD1_OPP5, RATE_IN_343X},
 };
 
 static struct vdd_prcm_config vdd2_rate_table[MAX_VDD2_OPP +1] = {
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/constraint.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/constraint.c	2011-01-11 23:47:20.000000000 -0500
@@ -66,7 +66,7 @@
 static struct atomic_notifier_head freq_dsp_pre_notifier_list;
 static struct atomic_notifier_head freq_dsp_post_notifier_list;
 
-#define min_vdd1_opp    CO_VDD1_OPP1
+#define min_vdd1_opp    CO_VDD1_OPP3
 #define max_vdd1_opp    CO_VDD1_OPP5
 #define min_vdd2_opp    CO_VDD2_OPP2
 #define max_vdd2_opp    CO_VDD2_OPP3
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/Kconfig	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24-AV8B/arch/arm/mach-omap3pe/Kconfig	2010-12-11 11:23:32.000000000 -0500
@@ -138,7 +138,7 @@
 	  Select this option if you want to boot at OPP4.
 
 config OMAP3ES2_VDD1_OPP5
-	bool "MPU-625Mhz/DSP-430Mhz(OPP5)"
+	bool "MPU-1005Mhz/DSP-430Mhz(OPP5)"
 	depends on OMAP3_PM
 	help
 	  Select this option if you want to boot at OPP5.
--- linux-2.6.24-palm/include/asm-arm/arch-omap/resource.h	2010-06-11 17:48:35.000000000 -0400
+++ linux-2.6.24-AV8B/include/asm-arm/arch-omap/resource.h	2011-02-19 10:35:23.000000000 -0500
@@ -29,14 +29,14 @@
 #include <linux/notifier.h>
 
 
-#define curr_vdd1_opp   3
+#define curr_vdd1_opp   5
 #define curr_vdd2_opp   3
-#define min_vdd1_opp    CO_VDD1_OPP1
+#define min_vdd1_opp    CO_VDD1_OPP3
 #define max_vdd1_opp    CO_VDD1_OPP5
 #define min_vdd2_opp    CO_VDD2_OPP2
 #define max_vdd2_opp    CO_VDD2_OPP3
 
-#define CURR_VDD1_OPP	3
+#define CURR_VDD1_OPP	5
 #define CURR_VDD2_OPP	3
 /* Flags to denote Pool usage */
 #define RES_UNUSED		0x0
@@ -269,15 +269,15 @@
 #define T2_VUSB3V1_MAX  0x2
 
 /* Frequencies and Voltages for OPPs */
-#define CURR_ARM_FREQ 	500
+#define CURR_ARM_FREQ 	1005
 #define CURR_DSP_FREQ 	360
-#define curr_arm_freq 	500
+#define curr_arm_freq 	1005
 #define curr_dsp_freq 	360
-#define min_arm_freq 	125
-#define max_arm_freq 	600
+#define min_arm_freq 	500
+#define max_arm_freq 	1005
 #define min_dsp_freq 	90
-#define max_dsp_freq 	430
-#define MIN_VDD1_OPP 	CO_VDD1_OPP1
+#define max_dsp_freq 	520
+#define MIN_VDD1_OPP 	CO_VDD1_OPP3
 #define MAX_VDD1_OPP 	CO_VDD1_OPP5
 #define MIN_VDD2_OPP 	CO_VDD2_OPP2
 #define MAX_VDD2_OPP 	CO_VDD2_OPP3
--- linux-2.6.24-palm/drivers/cpufreq/freq_table.c	2008-01-24 17:58:37.000000000 -0500
+++ linux-2.6.24-AV8B/drivers/cpufreq/freq_table.c	2010-12-11 13:04:48.000000000 -0500
@@ -186,7 +186,8 @@
 	table = show_table[cpu];
 
 	for (i=0; (table[i].frequency != CPUFREQ_TABLE_END); i++) {
-		if (table[i].frequency == CPUFREQ_ENTRY_INVALID)
+		if ((table[i].frequency == CPUFREQ_ENTRY_INVALID) ||
+		  (table[i].frequency < 1005000))
 			continue;
 		count += sprintf(&buf[count], "%d ", table[i].frequency);
 	}
--- linux-2.6.24-palm/include/linux/kernel.h	2008-01-24 17:58:37.000000000 -0500
+++ linux-2.6.24-AV8B/include/linux/kernel.h	2011-02-20 17:55:17.000000000 -0500
@@ -104,6 +104,7 @@
  * be bitten later when the calling function happens to sleep when it is not
  * supposed to.
  */
+#define CONFIG_PREEMPT_VOLUNTARY
 #ifdef CONFIG_PREEMPT_VOLUNTARY
 extern int cond_resched(void);
 # define might_resched() cond_resched()
--- linux-2.6.24-palm/block/deadline-iosched.c	2008-01-24 17:58:37.000000000 -0500
+++ linux-2.6.24-AV8B/block/deadline-iosched.c	2011-02-20 17:54:59.000000000 -0500
@@ -20,7 +20,7 @@
 static const int read_expire = HZ / 2;  /* max time before a read is submitted. */
 static const int write_expire = 5 * HZ; /* ditto for writes, these limits are SOFT! */
 static const int writes_starved = 2;    /* max times reads can starve a write */
-static const int fifo_batch = 16;       /* # of sequential requests treated as one
+static const int fifo_batch = 1;       /* # of sequential requests treated as one
 				     by the above parameters. For throughput. */
 
 struct deadline_data {
