--- a/Makefile	2011-11-06 11:44:37.000000000 -0500
+++ b/Makefile	2011-11-06 12:09:58.000000000 -0500
@@ -336,8 +336,8 @@
 CFLAGS_MODULE   = $(MODFLAGS)
 AFLAGS_MODULE   = $(MODFLAGS)
 LDFLAGS_MODULE  =
-CFLAGS_KERNEL	=
-AFLAGS_KERNEL	=
+CFLAGS_KERNEL	= -fgcse-lm -fgcse-sm -fsched-spec-load -fforce-addr -ffast-math -fsingle-precision-constant -mtune=cortex-a8 -mfpu=neon -ftree-vectorize -funswitch-loops
+AFLAGS_KERNEL	= -fgcse-lm -fgcse-sm -fsched-spec-load -fforce-addr -ffast-math -fsingle-precision-constant -mtune=cortex-a8 -mfpu=neon -ftree-vectorize -funswitch-loops
 
 
 # Use LINUXINCLUDE when you must reference the include/ directory.
