--- a/arch/arm/Kconfig	2011-07-30 12:55:51.000000000 -0400
+++ b/arch/arm/Kconfig	2011-07-31 15:15:31.000000000 -0400
@@ -1078,6 +1078,16 @@
           Scorpion processor supported hardware performance counters on a per
           thread basis or AXI counters on an overall system basis.
 
+config ALLOW_CPU_ALIGNMENT
+	bool "Allow CPU-based alignment handling"
+	default y
+	help
+	  Advanced ARM processors, such as the Cortex series and ARMv7-based
+	  CPUS are capable of performing unaligned accesses for many types of
+	  memory accesses.  Typically, using a cpu-based alignment fixup is
+	  faster than doing such a fixup in software.  For best performance
+	  on advanced CPUs, say Y here.
+
 endmenu
 
 source "arch/arm/common/Kconfig"
--- a/arch/arm/kernel/head-nommu.S	2010-08-01 18:11:14.000000000 -0400
+++ b/arch/arm/kernel/head-nommu.S	2011-07-31 15:31:20.000000000 -0400
@@ -65,10 +65,10 @@
 	 * CP15 system control register value returned in r0 from
 	 * the CPU init function.
 	 */
-#ifdef CONFIG_ALIGNMENT_TRAP
-	orr	r0, r0, #CR_A
-#else
+#ifdef CONFIG_ALLOW_CPU_ALIGNMENT
 	bic	r0, r0, #CR_A
+#else
+	orr	r0, r0, #CR_A
 #endif
 #ifdef CONFIG_CPU_DCACHE_DISABLE
 	bic	r0, r0, #CR_C
--- a/arch/arm/kernel/head.S	2011-07-30 12:55:51.000000000 -0400
+++ b/arch/arm/kernel/head.S	2011-07-31 15:15:31.000000000 -0400
@@ -160,10 +160,10 @@
  * registers.
  */
 __enable_mmu:
-#ifdef CONFIG_ALIGNMENT_TRAP
-	orr	r0, r0, #CR_A
-#else
+#ifdef CONFIG_ALLOW_CPU_ALIGNMENT
 	bic	r0, r0, #CR_A
+#else
+	orr	r0, r0, #CR_A
 #endif
 #ifdef CONFIG_CPU_DCACHE_DISABLE
 	bic	r0, r0, #CR_C
