--- linux-2.6.24-palm/arch/arm/mach-omap3pe/clock.h	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24.7-F105/arch/arm/mach-omap3pe/clock.h	2010-11-02 18:23:45.000000000 -0400
@@ -41,15 +41,16 @@
 #define S38M	38400000
 #define S96M	96000000
 #define S83M	83000000
-#define S166M	166000000
+#define S200M	200000000
 
 #define S120M	120000000
 #define S125M	125000000
 #define S250M	250000000
 #define S500M	500000000
-#define S550M	550000000
-#define S625M	625000000
-#define S600M	600000000
+#define S700M	700000000
+#define S800M	800000000
+#define S900M	900000000
+#define S1005M	1005000000
 
 /* Macro to enable clock control via clock framework */
 #define ENABLE_CLOCKCONTROL 1
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/clock_tree.h	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24.7-F105/arch/arm/mach-omap3pe/clock_tree.h	2010-11-02 18:23:45.000000000 -0400
@@ -90,9 +90,14 @@
 	/*OPP3*/
 	{S500M, PRCM_VDD1_OPP3, RATE_IN_343X},
 	/*OPP4*/
-	{S550M, PRCM_VDD1_OPP4, RATE_IN_343X},
+	{S700M, PRCM_VDD1_OPP4, RATE_IN_343X},
 	/*OPP5*/
-	{S600M, PRCM_VDD1_OPP5, RATE_IN_343X},
+	{S800M, PRCM_VDD1_OPP5, RATE_IN_343X},
+        /*OPP6*/
+        {S900M, PRCM_VDD1_OPP6, RATE_IN_343X},
+        /*OPP7*/
+        {S1005M, PRCM_VDD1_OPP7, RATE_IN_343X},
+
 };
 
 static struct vdd_prcm_config vdd2_rate_table[MAX_VDD2_OPP +1] = {
@@ -102,7 +107,7 @@
 	/*OPP2*/
 	{S83M, PRCM_VDD2_OPP2, RATE_IN_343X},
 	/*OPP3*/
-	{S166M, PRCM_VDD2_OPP3, RATE_IN_343X},
+	{S200M, PRCM_VDD2_OPP3, RATE_IN_343X},
 };
 
 /* Base external input clocks */
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/constraint.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24.7-F105/arch/arm/mach-omap3pe/constraint.c	2010-12-23 10:04:10.000000000 -0500
@@ -67,7 +67,7 @@
 static struct atomic_notifier_head freq_dsp_post_notifier_list;
 
 #define min_vdd1_opp    CO_VDD1_OPP1
-#define max_vdd1_opp    CO_VDD1_OPP5
+#define max_vdd1_opp    CO_VDD1_OPP7
 #define min_vdd2_opp    CO_VDD2_OPP2
 #define max_vdd2_opp    CO_VDD2_OPP3
 
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/Kconfig	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24.7-F105/arch/arm/mach-omap3pe/Kconfig	2010-11-02 18:23:45.000000000 -0400
@@ -132,17 +132,29 @@
 	  Select this option if you want to boot at OPP3.
 
 config OMAP3ES2_VDD1_OPP4
-	bool "MPU-550Mhz/DSP-396Mhz(OPP4)"
+	bool "MPU-700Mhz/DSP-360Mhz(OPP4)"
 	depends on OMAP3_PM
 	help
 	  Select this option if you want to boot at OPP4.
 
 config OMAP3ES2_VDD1_OPP5
-	bool "MPU-625Mhz/DSP-430Mhz(OPP5)"
+	bool "MPU-800Mhz/DSP-360Mhz(OPP5)"
 	depends on OMAP3_PM
 	help
 	  Select this option if you want to boot at OPP5.
 
+config OMAP3ES2_VDD1_OPP6
+        bool "MPU-900Mhz/DSP-360Mhz(OPP6)"
+        depends on OMAP3_PM
+        help
+          Select this option if you want to boot at OPP6.
+
+config OMAP3ES2_VDD1_OPP7
+        bool "MPU-1005Mhz/DSP-360Mhz(OPP7)"
+        depends on OMAP3_PM
+        help
+          Select this option if you want to boot at OPP7.
+
 endchoice
 
 comment "Default VDD2 OPP Level Selection"
--- linux-2.6.24-palm/arch/arm/mach-omap3pe/prcm_opp.c	2010-06-11 17:48:34.000000000 -0400
+++ linux-2.6.24.7-F105/arch/arm/mach-omap3pe/prcm_opp.c	2010-12-24 14:49:13.000000000 -0500
@@ -56,13 +56,13 @@
 
 static u8 mpu_iva2_vdd1_volts [PRCM_NO_VDD1_OPPS] = {
 	/* Vsel corresponding to 0.975V (OPP1), 1.050V (OPP2),
-				1.20V (OPP3), 1.27V (OPP4), 1.35 (OPP5) */
-	0x1e, 0x24, 0x30, 0x36, 0x3C
+	1.20V (OPP3), 1.27V (OPP4), 1.42V (OPP5), 1.45V (OPP6), 1.46V (OPP7) */
+		0x1E, 0x24, 0x28, 0x3f, 0x40, 0x42, 0x43
 };
 
 static u8 core_l3_vdd2_volts [PRCM_NO_VDD2_OPPS] = { /* only 3 OPPs */
 	/* Vsel corresponding to 0.975V (OPP1), 1.05V (OPP2), 1.15 (OPP3) */
-	0x1e, 0x24, 0x2c
+	0x1e, 0x24, 0x2e
 };
 
 
@@ -82,7 +82,7 @@
 	 {0x00101001, 0x00040402, 0x0E050E05, 0x020F1010, 0x0F0502C2} }
 	},
 
-	/* SDRC CS0/CS1 values 166MHZ*/
+	/* SDRC CS0/CS1 values 200MHZ*/
 	{
 	{{0x0004e201, 0x629db4c6, 0x00012214},
 	 {0x0004e201, 0x629db4c6, 0x00012214} },
@@ -155,20 +155,32 @@
 	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
 	{0x0FA, 0x05, 0x07, 0x01}, {0x113, 0x05, 0x07, 0x01},
 	/* OPP5 (600 Mhz) */
+	{0x12C, 0x05, 0x07, 0x01},
+	/* OPP6 (720 Mhz) */
+	{0x12C, 0x05, 0x07, 0x01},
+	/* OPP7 (800 Mhz) */
 	{0x12C, 0x05, 0x07, 0x01} },
 	/* 13M values */
 	/* OPP1(125 Mhz) and OPP2(250 Mhz)*/
 	{{0x1F4, 0x0C, 0x03, 0x04}, {0x1F4, 0x0C, 0x03, 0x02},
-	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
-	{0x1F4, 0x0C, 0x03, 0x01}, {0x226, 0x0C, 0x03, 0x01},
-	/* OPP5 (600 Mhz) */
-	{0x258, 0x0C, 0x03, 0x01} },
+	/* OPP3(500 Mhz) and OPP4(700 Mhz)*/
+	{0x1F4, 0x0C, 0x03, 0x01}, {0x2BC, 0x0C, 0x03, 0x01},
+	/* OPP5 (800 Mhz) */
+	{0x320, 0x0C, 0x03, 0x01},
+	/* OPP6 (900 Mhz) */
+	{0x384, 0x0C, 0x03, 0x01},
+	/* OPP7 (1005 Mhz) */
+	{0x3ED, 0x0C, 0x03, 0x01} },
 	/* 19.2M values */
 	/* OPP1(125 Mhz) and OPP2(250 Mhz)*/
 	{{0x271, 0x17, 0x03, 0x04}, {0x271, 0x17, 0x03, 0x02},
 	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
 	{0x271, 0x17, 0x03, 0x01}, {0x191, 0x0D, 0x05, 0x01},
 	/* OPP5 (600 Mhz) */
+	{0x177, 0x0B, 0x06, 0x01},
+	/* OPP6 (720 Mhz) */
+	{0x177, 0x0B, 0x06, 0x01},
+	/* OPP7 (800 Mhz) */
 	{0x177, 0x0B, 0x06, 0x01} },
 	/* 26M values */
 	/* OPP1(125 Mhz) and OPP2(250 Mhz)*/
@@ -176,13 +188,21 @@
 	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
 	{0x0FA, 0x0C, 0x07, 0x01}, {0x113, 0x0C, 0x07, 0x01},
 	/* OPP5 (600 Mhz) */
-	{0x12C, 0x0C, 0x07, 0x01} },
+	{0x12C, 0x0C, 0x07, 0x01},
+	/* OPP6 (720 Mhz) */
+	{0x168, 0x0C, 0x07, 0x01},
+	/* OPP7 (800 Mhz) */
+	{0x190, 0x0C, 0x07, 0x01} },
 	/* 38.4M values */
 	/* OPP1(125 Mhz) and OPP2(250 Mhz)*/
 	{{0x271, 0x2F, 0x03, 0x04}, {0x271, 0x2F, 0x03, 0x02},
 	/* OPP3(500 Mhz) and OPP4(550 Mhz)*/
 	{0x271, 0x2F, 0x03, 0x01}, {0x1BC, 0x1E, 0x04, 0x01},
 	/* OPP5 (600 Mhz) */
+	{0x177, 0x17, 0x06, 0x01},
+	/* OPP6 (720 Mhz) */
+	{0x177, 0x17, 0x06, 0x01},
+	/* OPP7 (800 Mhz) */
 	{0x177, 0x17, 0x06, 0x01} },
 };
 
@@ -194,7 +214,11 @@
 	/* OPP3(360 Mhz) and OPP4(396 Mhz)*/
 	 {0x0B4, 0x05, 0x07, 0x01}, {0x0C6, 0x05, 0x07, 0x01},
 	/* OPP5 (430 Mhz) */
-	 {0x0D7, 0x05, 0x07, 0x01} },
+	 {0x0D7, 0x05, 0x07, 0x01},
+        /* OPP6 (520 Mhz) */
+        {0x0D7, 0x05, 0x07, 0x01},
+        /* OPP7 (600 Mhz) */
+        {0x0D7, 0x05, 0x07, 0x01} },
 #if 0
 	/* 13M values */
 	/* OPP1(90 Mhz) and OPP2(180 Mhz)*/
@@ -215,6 +239,10 @@
 	/* OPP3(360 Mhz) and OPP4(360 Mhz)*/
 	 {0x168, 0x0C, 0x03, 0x01}, {0x168, 0x0C, 0x03, 0x01},
 	/* OPP5 (360 Mhz) */
+	 {0x168, 0x0C, 0x03, 0x01},
+        /* OPP6 (360 Mhz) */
+	 {0x168, 0x0C, 0x03, 0x01},
+        /* OPP7 (360 Mhz) */
 	 {0x168, 0x0C, 0x03, 0x01} },
 #endif
 	/* 19.2M values */
@@ -223,31 +251,43 @@
 	/* OPP3(360 Mhz) and OPP4(396 Mhz)*/
 	 {0x0E1, 0x0B, 0x06, 0x01}, {0x14A, 0x0F, 0x04, 0x01},
 	/* OPP5 (430 Mhz) */
-	 {0x203, 0x16, 0x03, 0x01} },
+	 {0x203, 0x16, 0x03, 0x01},
+        /* OPP6 (520 Mhz) */
+        {0x203, 0x16, 0x03, 0x01},
+        /* OPP7 (600 Mhz) */
+        {0x203, 0x16, 0x03, 0x01} },
 	/* 26M values */
 	/* OPP1(90 Mhz) and OPP2(180 Mhz)*/
 	{{0x0B4, 0x0C, 0x07, 0x04}, {0x0B4, 0x0C, 0x07, 0x02},
 	/* OPP3(360 Mhz) and OPP4(396 Mhz)*/
 	 {0x0B4, 0x0C, 0x07, 0x01}, {0x0C6, 0x0C, 0x07, 0x01},
 	/* OPP5 (430 Mhz) */
-	 {0x0D7, 0x0C, 0x07, 0x01} },
+	 {0x0D7, 0x0C, 0x07, 0x01},
+        /* OPP6 (520 Mhz) */
+        {0x104, 0x0C, 0x07, 0x01},
+        /* OPP7 (600 Mhz) */
+        {0x12c, 0x0C, 0x07, 0x01} },
 	/* 38.4M values */
 	/* OPP1(90 Mhz) and OPP2(180 Mhz)*/
 	{{0x0E1, 0x17, 0x06, 0x04}, {0x0E1, 0x17, 0x06, 0x02},
 	/* OPP3(360 Mhz) and OPP4(396 Mhz)*/
 	 {0x0E1, 0x17, 0x06, 0x01}, {0x14A, 0x1F, 0x04, 0x01},
 	/* OPP5 (430 Mhz) */
-	 {0x23B, 0x32, 0x01, 0x01} },
+	 {0x23B, 0x32, 0x01, 0x01},
+        /* OPP6 (520 Mhz) */
+        {0x23B, 0x32, 0x03, 0x01},
+        /* OPP7 (600 Mhz) */
+        {0x23B, 0x32, 0x03, 0x01} },
 };
 
 /* CORE parameters */
 static struct dpll_param core_dpll_param[5][PRCM_NO_VDD2_OPPS] = {
 	/* 12M values */
-	/* OPP1(83 Mhz) and OPP2(166 Mhz) */
+	/* OPP1(83 Mhz) and OPP2(200 Mhz) */
 	{{0, 0, 0, 0}, {0x0A6, 0x05, 0x07, 0x02}, {0x0A6, 0x05, 0x07, 0x01} },
 	/* 13M values */
-	/* OPP1(83 Mhz) and OPP2(166 Mhz) */
-	{{0, 0, 0, 0}, {0x14C, 0x0C, 0x03, 0x02}, {0x14C, 0x0C, 0x03, 0x01} },
+	/* OPP1(83 Mhz) and OPP2(200 Mhz) */
+	{{0, 0, 0, 0}, {0x14C, 0x0C, 0x03, 0x02}, {0x190, 0x0C, 0x03, 0x01} },
 	/* 19.2M values */
 	/* OPP1(83 Mhz) and OPP2(166 Mhz) */
 	{{0, 0, 0, 0}, {0x19F, 0x17, 0x03, 0x02}, {0x19F, 0x17, 0x03, 0x01} },
@@ -459,7 +499,7 @@
 	/* Scale back to previous OPP for VDD2*/
 	if (current_vdd2_opp != PRCM_VDD2_OPP2) {
 		prcm_do_voltage_scaling(PRCM_VDD2_OPP3, PRCM_VDD2_OPP2);
-		valid_rate = clk_round_rate(p_vdd2_clk, S166M);
+		valid_rate = clk_round_rate(p_vdd2_clk, S200M);
 		p_vdd2_clk->set_rate(p_vdd2_clk, valid_rate);
 	}
 
@@ -473,11 +513,18 @@
 			valid_rate = clk_round_rate(p_vdd1_clk, S500M);
 			break;
 		case PRCM_VDD1_OPP4:
-			valid_rate = clk_round_rate(p_vdd1_clk, S550M);
+			valid_rate = clk_round_rate(p_vdd1_clk, S700M);
 			break;
 		case PRCM_VDD1_OPP5:
-			valid_rate = clk_round_rate(p_vdd1_clk, S600M);
+			valid_rate = clk_round_rate(p_vdd1_clk, S800M);
 			break;
+                case PRCM_VDD1_OPP6:
+                        valid_rate = clk_round_rate(p_vdd1_clk, S900M);
+                        break;
+                case PRCM_VDD1_OPP7:
+                        valid_rate = clk_round_rate(p_vdd1_clk, S1005M);
+                        break;
+
 		}
 		p_vdd1_clk->set_rate(p_vdd1_clk, valid_rate);
 	}
@@ -490,11 +537,14 @@
 	unsigned int constraint;
 	unsigned int opp;
 } vdd1_arm_dsp_freq[MAX_VDD1_OPP] = {
-	{125,  90, CO_VDD1_OPP1, PRCM_VDD1_OPP1},
-	{250, 180, CO_VDD1_OPP2, PRCM_VDD1_OPP2},
+	{125, 90, CO_VDD1_OPP1, PRCM_VDD1_OPP1},
+	{250, 90, CO_VDD1_OPP2, PRCM_VDD1_OPP2},
 	{500, 360, CO_VDD1_OPP3, PRCM_VDD1_OPP3},
-	{550, 396, CO_VDD1_OPP4, PRCM_VDD1_OPP4},
-	{600, 430, CO_VDD1_OPP5, PRCM_VDD1_OPP5},
+	{700, 360, CO_VDD1_OPP4, PRCM_VDD1_OPP4},
+	{800, 360, CO_VDD1_OPP5, PRCM_VDD1_OPP5},
+	{900, 360, CO_VDD1_OPP6, PRCM_VDD1_OPP6},
+	{1005, 360, CO_VDD1_OPP7, PRCM_VDD1_OPP7},
+
 };
 static struct vdd2_core_freq_d {
 	unsigned int freq;
@@ -503,14 +553,14 @@
 } vdd2_core_freq[MAX_VDD2_OPP] = {
 	{0, CO_VDD2_OPP1, PRCM_VDD2_OPP1},
 	{83, CO_VDD2_OPP2, PRCM_VDD2_OPP2},
-	{166, CO_VDD2_OPP3, PRCM_VDD2_OPP3},
+	{200, CO_VDD2_OPP3, PRCM_VDD2_OPP3},
 };
 
-static unsigned int rnd_rate_vdd1[5] = {
-	S125M, S250M, S500M, S550M, S600M
+static unsigned int rnd_rate_vdd1[7] = {
+	S125M, S250M, S500M, S700M, S800M, S900M, S1005M
 };
 static unsigned int rnd_rate_vdd2[3] = {
-	0, S83M, S166M
+	0, S83M, S200M
 };
 
 /* To set the opp for vdd1 */
@@ -1449,6 +1499,11 @@
 	target_vdd1_opp = PRCM_VDD1_OPP4;
 #elif defined(CONFIG_OMAP3ES2_VDD1_OPP5)
 	target_vdd1_opp = PRCM_VDD1_OPP5;
+#elif defined(CONFIG_OMAP3ES2_VDD1_OPP6)
+        target_vdd1_opp = PRCM_VDD1_OPP6;
+#elif defined(CONFIG_OMAP3ES2_VDD1_OPP7)
+        target_vdd1_opp = PRCM_VDD1_OPP7;
+
 #endif
         target_mpu_khz  = get_arm_freq_for_opp(get_opp_no(target_vdd1_opp));
 
--- linux-2.6.24-palm/include/asm-arm/arch-omap/prcm.h	2010-06-11 17:48:35.000000000 -0400
+++ linux-2.6.24.7-F105/include/asm-arm/arch-omap/prcm.h	2010-11-02 18:23:45.000000000 -0400
@@ -609,7 +609,12 @@
 							| ID_OPP_NO(0x4))
 #define PRCM_VDD1_OPP5		(OMAP(AT_3430)| OTHER_ID_TYPE(ID_OPP) | ID_VDD(PRCM_VDD1)\
 							| ID_OPP_NO(0x5))
-#define PRCM_NO_VDD1_OPPS	5
+#define PRCM_VDD1_OPP6          (OMAP(AT_3430)| OTHER_ID_TYPE(ID_OPP) | ID_VDD(PRCM_VDD1)\
+                                                        | ID_OPP_NO(0x6))
+#define PRCM_VDD1_OPP7          (OMAP(AT_3430)| OTHER_ID_TYPE(ID_OPP) | ID_VDD(PRCM_VDD1)\
+                                                        | ID_OPP_NO(0x7))
+
+#define PRCM_NO_VDD1_OPPS	7
 
 /* VDD2 OPPs */
 #define PRCM_VDD2_OPP1		(OMAP(AT_3430)| OTHER_ID_TYPE(ID_OPP) | ID_VDD(PRCM_VDD2)\
--- linux-2.6.24-palm/include/asm-arm/arch-omap/resource.h	2010-06-11 17:48:35.000000000 -0400
+++ linux-2.6.24.7-F105/include/asm-arm/arch-omap/resource.h	2010-12-23 10:04:00.000000000 -0500
@@ -32,7 +32,7 @@
 #define curr_vdd1_opp   3
 #define curr_vdd2_opp   3
 #define min_vdd1_opp    CO_VDD1_OPP1
-#define max_vdd1_opp    CO_VDD1_OPP5
+#define max_vdd1_opp    CO_VDD1_OPP7
 #define min_vdd2_opp    CO_VDD2_OPP2
 #define max_vdd2_opp    CO_VDD2_OPP3
 
@@ -121,6 +121,8 @@
 #define CO_VDD1_OPP3		0x3
 #define CO_VDD1_OPP4		0x4
 #define CO_VDD1_OPP5		0x5
+#define CO_VDD1_OPP6		0x6
+#define CO_VDD1_OPP7		0x7
 
 /* VDD2 OPP constraint levels */
 #define CO_VDD2_OPP1		0x1
@@ -273,12 +275,12 @@
 #define CURR_DSP_FREQ 	360
 #define curr_arm_freq 	500
 #define curr_dsp_freq 	360
-#define min_arm_freq 	125
-#define max_arm_freq 	600
-#define min_dsp_freq 	90
-#define max_dsp_freq 	430
+#define min_arm_freq 	500
+#define max_arm_freq 	1005
+#define min_dsp_freq 	360
+#define max_dsp_freq 	520
 #define MIN_VDD1_OPP 	CO_VDD1_OPP1
-#define MAX_VDD1_OPP 	CO_VDD1_OPP5
+#define MAX_VDD1_OPP 	CO_VDD1_OPP7
 #define MIN_VDD2_OPP 	CO_VDD2_OPP2
 #define MAX_VDD2_OPP 	CO_VDD2_OPP3
 
