--- a/include/linux/mm.h	2011-08-02 16:04:22.000000000 -0400
+++ b/include/linux/mm.h	2011-09-27 21:10:30.000000000 -0400
@@ -1308,7 +1308,7 @@
 void task_dirty_inc(struct task_struct *tsk);
 
 /* readahead.c */
-#define VM_MAX_READAHEAD	128	/* kbytes */
+#define VM_MAX_READAHEAD	1024	/* kbytes */
 #define VM_MIN_READAHEAD	16	/* kbytes (includes current page) */
 
 int force_page_cache_readahead(struct address_space *mapping, struct file *filp,
--- a/include/linux/blkdev.h	2010-08-01 18:11:14.000000000 -0400
+++ b/include/linux/blkdev.h	2011-09-27 21:23:39.000000000 -0400
@@ -32,7 +32,7 @@
 struct sg_io_hdr;
 
 #define BLKDEV_MIN_RQ	4
-#define BLKDEV_MAX_RQ	128	/* Default maximum */
+#define BLKDEV_MAX_RQ	1024	/* Default maximum */
 
 struct request;
 typedef void (rq_end_io_fn)(struct request *, int);
