diff --git a/drivers/input/keyboard/maxim7359-keypad.c b/drivers/input/keyboard/maxim7359-keypad.c
index 7063bd8..a93c56a 100644
--- a/drivers/input/keyboard/maxim7359-keypad.c
+++ b/drivers/input/keyboard/maxim7359-keypad.c
@@ -381,6 +381,35 @@ static DEVICE_ATTR(fifo_read, S_IRUGO, fifo_read_show, NULL);
 
 
 
+static ssize_t
+prox_timeout_show(	struct device *dev,
+		struct device_attribute *attr,
+		char *buf)
+{
+	struct maxim_kp_state *state = (struct maxim_kp_state *)dev_get_drvdata(dev);
+
+	return snprintf(buf, PAGE_SIZE, "%d\n", jiffies_to_msecs(state->prox_timeout));
+}
+
+static ssize_t
+prox_timeout_store( struct device *dev,
+                struct device_attribute *attr,
+                const char *buf, size_t count)
+{
+	char *endp;
+	struct maxim_kp_state *state = (struct maxim_kp_state *)dev_get_drvdata(dev);
+	int new_prox_timeout = simple_strtoul(buf, &endp, 10);
+
+  if (new_prox_timeout)
+	  state->prox_timeout = msecs_to_jiffies(state->pdata->key_prox_timeout); 
+  else
+	  state->prox_timeout = 0;
+
+	return count;
+}
+
+static DEVICE_ATTR(prox_timeout, S_IWUGO | S_IRUGO, prox_timeout_show, prox_timeout_store);
+
 /******************************************************************************
  *****************************************************************************/
 
@@ -701,6 +730,11 @@ static int maxim_i2c_probe(struct i2c_client *client)
 		goto err2;
 	}
 
+	rc = device_create_file(&(state->i2c_dev->dev), &dev_attr_prox_timeout);
+	if(rc) {
+		goto err2;
+	}
+
 #ifdef MAXIM7359_DEBUG
 	rc = device_create_file(&(state->i2c_dev->dev),  &dev_attr_hw_debounce);
 	if(rc) {
@@ -812,6 +846,7 @@ static int maxim_i2c_remove(struct i2c_client *client)
 
 	device_remove_file(&(state->i2c_dev->dev), &dev_attr_key_state);
 	device_remove_file(&(state->i2c_dev->dev), &dev_attr_fifo_read);
+	device_remove_file(&(state->i2c_dev->dev), &dev_attr_prox_timeout);
 #ifdef MAXIM7359_DEBUG
 	device_remove_file(&(state->i2c_dev->dev), &dev_attr_sw_debounce);
 	device_remove_file(&(state->i2c_dev->dev), &dev_attr_hw_debounce);
