diff -c /srv/preware/build/testing-kernels/uber-kernel-pre/build/src-1.4.5-58/linux-2.6.24/arch/arm/mach-omap3pe/board-nduid.c\~ /srv/preware/build/testing-kernels/uber-kernel-pre/build/src-1.4.5-58/linux-2.6.24/arch/arm/mach-omap3pe/board-nduid.c
*** /srv/preware/build/testing-kernels/uber-kernel-pre/build/src-1.4.5-58/linux-2.6.24/arch/arm/mach-omap3pe/board-nduid.c~	2010-07-12 12:24:36.763254218 +0930
--- /srv/preware/build/testing-kernels/uber-kernel-pre/build/src-1.4.5-58/linux-2.6.24/arch/arm/mach-omap3pe/board-nduid.c	2010-07-12 12:40:39.800754358 +0930
***************
*** 18,23 ****
--- 18,24 ----
   * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
   */
  
+ #include <linux/kernel.h>
  #include <asm/arch/nduid.h>
  #include <asm/io.h>
  #if defined(CONFIG_ARCH_OMAP24XX)
***************
*** 63,67 ****
--- 64,72 ----
  	buf[4] = __raw_readl(CONTROL_TAP_DIE_ID_2);
  	buf[5] = __raw_readl(CONTROL_TAP_DIE_ID_3);
  
+ 	printk(KERN_INFO "IDCODE:  0x%04x\n", buf[0]);
+ 	printk(KERN_INFO "PROD_ID: 0x%04x\n", buf[1]);
+ 	printk(KERN_INFO "DIE_ID:  0x%04x%04x%04x%04x\n", buf[2], buf[3], buf[4], buf[5]);
+ 
  	return sizeof(uint32_t) * 6;
  }

Diff finished.  Mon Jul 12 12:41:50 2010
