--- drivers/hwmon/omap34xx_temp.c	2010-07-04 22:27:34.000000000 -0700
+++ drivers/hwmon/omap36xx_temp.c	2010-11-24 18:15:50.000000000 -0800
@@ -34,8 +34,8 @@
 #include <asm/arch/omap34xx.h>
 #include <asm/arch/control.h>
 
-#define TEMP_SENSOR_SOC BIT(8)
-#define TEMP_SENSOR_EOCZ BIT(7)
+#define TEMP_SENSOR_SOC BIT(9)
+#define TEMP_SENSOR_EOCZ BIT(8)
 
 /* minimum delay for EOCZ rise after SOC rise is
  * 11 cycles of the 32.768Khz clock */
@@ -132,7 +132,7 @@
 			goto err;
 		}
 		data->temp = omap_ctrl_readl(OMAP343X_CONTROL_TEMP_SENSOR) &
-						((1<<7) - 1);
+						((1<<8) - 1);
 		omap34xx_temp_cooked = data->temp ;
  		data->last_updated = jiffies;
 		data->valid = 1;
