--- drivers/hwmon/omap34xx_temp_orig.c	2010-07-04 22:06:00.000000000 -0700
+++ drivers/hwmon/omap34xx_temp.c	2010-07-04 22:19:24.000000000 -0700
@@ -4,6 +4,8 @@
  * Copyright (C) 2008 Nokia Corporation
  *
  * Written by Peter De Schrijver <peter.de-schrijver@nokia.com>
+ * 
+ * Modified by Tom King <ka6sox#gmail.com>
  *
  * Inspired by k8temp.c
  *
@@ -62,6 +64,7 @@
 	u32 temp;
 };
 
+static u32 omap34xx_temp_cooked;
 
 static struct platform_device omap34xx_temp_device = {
 	.name 	= "omap34xx_temp",
@@ -130,7 +133,8 @@
 		}
 		data->temp = omap_ctrl_readl(OMAP343X_CONTROL_TEMP_SENSOR) &
 						((1<<7) - 1);
-		data->last_updated = jiffies;
+		omap34xx_temp_cooked = data->temp ;
+ 		data->last_updated = jiffies;
 		data->valid = 1;
 
 err:
@@ -225,6 +229,8 @@
 
 	return 0;
 
+EXPORT_SYMBOL(omap34xx_temp_cooked);
+
 exit_remove_all:
 	device_remove_file(&omap34xx_temp_device.dev,
 			   &dev_attr_name);
--- drivers/hwmon/omap34xx_temp_orig.c	2010-07-04 22:06:00.000000000 -0700
+++ drivers/hwmon/omap34xx_temp.c	2010-07-04 22:27:34.000000000 -0700
@@ -4,6 +4,8 @@
  * Copyright (C) 2008 Nokia Corporation
  *
  * Written by Peter De Schrijver <peter.de-schrijver@nokia.com>
+ * 
+ * Modified by Tom King <ka6sox#gmail.com>
  *
  * Inspired by k8temp.c
  *
@@ -62,6 +64,7 @@
 	u32 temp;
 };
 
+static u32 omap34xx_temp_cooked;
 
 static struct platform_device omap34xx_temp_device = {
 	.name 	= "omap34xx_temp",
@@ -130,7 +133,8 @@
 		}
 		data->temp = omap_ctrl_readl(OMAP343X_CONTROL_TEMP_SENSOR) &
 						((1<<7) - 1);
-		data->last_updated = jiffies;
+		omap34xx_temp_cooked = data->temp ;
+ 		data->last_updated = jiffies;
 		data->valid = 1;
 
 err:
@@ -174,6 +178,8 @@
 				NULL, 0, 0);
 static DEVICE_ATTR(name, S_IRUGO, show_name, NULL);
 
+EXPORT_SYMBOL(omap34xx_temp_cooked);
+
 static int __devinit omap34xx_temp_probe(void)
 {
 	int err;
