*** linux-2.6.24/arch/arm/mach-msm/cpufreq.c~	2010-10-13 19:26:37.000000000 +1030
--- linux-2.6.24/arch/arm/mach-msm/cpufreq.c	2010-10-25 23:04:48.000000000 +1030
***************
*** 109,114 ****
--- 109,121 ----
  	return 0;
  }
  
+ #ifdef CONFIG_CPU_FREQ_TABLE
+ static struct freq_attr *msm_cpufreq_attr[] = {
+ 	&cpufreq_freq_attr_scaling_available_freqs,
+ 	NULL,
+ };
+ #endif
+ 
  static struct cpufreq_driver msm_cpufreq_driver = {
  	/* lps calculations are handled here. */
  	.flags		= CPUFREQ_STICKY | CPUFREQ_CONST_LOOPS,
***************
*** 116,121 ****
--- 123,131 ----
  	.verify		= msm_cpufreq_verify,
  	.target		= msm_cpufreq_target,
  	.name		= "msm",
+ #ifdef CONFIG_CPU_FREQ_TABLE
+ 	.attr		= msm_cpufreq_attr,
+ #endif
  };
  
  static int __init msm_cpufreq_register(void)

