*** linux-2.6.24/arch/arm/configs/chuck_defconfig~	2010-10-25 23:24:51.000000000 +1030
--- linux-2.6.24/arch/arm/configs/chuck_defconfig	2010-10-25 23:26:34.000000000 +1030
***************
*** 187,193 ****
  CONFIG_MSM_ONCRPCROUTER=y
  # CONFIG_MSM_ONCRPCROUTER_DEBUG is not set
  CONFIG_MSM_RPCSERVERS=y
! # CONFIG_MSM_CPU_FREQ_SET_MIN_MAX is not set
  CONFIG_MSM_ADSP=y
  CONFIG_AUDIO_AAC_PLUS=y
  CONFIG_AUDIO_ENHANCED_AAC_PLUS=y
--- 187,195 ----
  CONFIG_MSM_ONCRPCROUTER=y
  # CONFIG_MSM_ONCRPCROUTER_DEBUG is not set
  CONFIG_MSM_RPCSERVERS=y
! CONFIG_MSM_CPU_FREQ_SET_MIN_MAX=y
! CONFIG_MSM_CPU_FREQ_MIN=122800
! CONFIG_MSM_CPU_FREQ_MAX=600000
  CONFIG_MSM_ADSP=y
  CONFIG_AUDIO_AAC_PLUS=y
  CONFIG_AUDIO_ENHANCED_AAC_PLUS=y
***************
*** 281,287 ****
  #
  CONFIG_CPU_FREQ=y
  CONFIG_CPU_FREQ_TABLE=y
! # CONFIG_CPU_FREQ_DEBUG is not set
  CONFIG_CPU_FREQ_STAT=y
  CONFIG_CPU_FREQ_STAT_DETAILS=y
  # CONFIG_CPU_FREQ_DEFAULT_GOV_PERFORMANCE is not set
--- 283,289 ----
  #
  CONFIG_CPU_FREQ=y
  CONFIG_CPU_FREQ_TABLE=y
! CONFIG_CPU_FREQ_DEBUG=y
  CONFIG_CPU_FREQ_STAT=y
  CONFIG_CPU_FREQ_STAT_DETAILS=y
  # CONFIG_CPU_FREQ_DEFAULT_GOV_PERFORMANCE is not set
*** linux-2.6.24/drivers/cpufreq/cpufreq.c~	2010-10-13 19:26:39.000000000 +1030
--- linux-2.6.24/drivers/cpufreq/cpufreq.c	2010-10-25 23:36:33.000000000 +1030
***************
*** 774,781 ****
--- 774,786 ----
  		unlock_policy_rwsem_write(cpu);
  		goto err_out;
  	}
+ #ifdef CONFIG_MSM_CPU_FREQ_SET_MIN_MAX
+ 	policy->user_policy.min = policy->min;
+ 	policy->user_policy.max = policy->max;
+ #else
  	policy->user_policy.min = policy->cpuinfo.min_freq;
  	policy->user_policy.max = policy->cpuinfo.max_freq;
+ #endif
  
  #ifdef CONFIG_SMP
  

