--- linux-2.6.24/drivers/cpufreq/cpufreq_override.c.orig	2010-08-01 21:31:12.000000000 +0930
+++ linux-2.6.24/drivers/cpufreq/cpufreq_override.c	2010-08-01 21:31:22.000000000 +0930
@@ -140,7 +140,7 @@
 	return count;
 }
 
-CPUFREQ_OVERRIDE_ATTR(vdd1_vsel,0644,show_vdd1_vsel,store_vdd1_vsel);
+CPUFREQ_OVERRIDE_ATTR(vdd1_vsel,0644,show_vdd1_vsel,NULL);
 CPUFREQ_OVERRIDE_ATTR(cpu_hightemp_alarm,0644,show_hightemp_scaleback,
 			store_hightemp_scaleback);
 CPUFREQ_OVERRIDE_ATTR(cpu_hightemp_reset,0644,show_lowtemp_reset,
